### PR TITLE
feat: Station search bar

### PR DESCRIPTION
**Pull request by cmaddox5**
**Asana task**: [Station search bar](https://app.asana.com/0/1185117109217413/1205855245091615/f)

Added a `SearchBar` component. This uses a third-party package that has basically everything we need for searching, autocompleting, and selecting results. There are a bunch of [props](https://github.com/sickdyd/react-search-autocomplete) to use that give us control over how the search works. For example, `fuseOptions` gives us the ability to adjust things like  `minMatchCharLength` (how many characters need to match before the result shows up) and `threshold` (how strict the search needs to be). Documentation for these options can be found [here](https://www.fusejs.io/).

### CODE REVIEW

**Comment by PaulJKim** [comment]
Screenshot 2023-12-11 at 10 20 18 AM

Not sure if this is a bug with the component itself, but I found that letters that I type in will occasionally be automatically deleted. In the above screenshot, I briefly saw the `e` in Medford but as I continue typing, it was deleted without me pressing backspace.

**Comment by cmaddox5** [comment]
> Screenshot 2023-12-11 at 10 20 18 AM
> 
> Not sure if this is a bug with the component itself, but I found that letters that I type in will occasionally be automatically deleted. In the above screenshot, I briefly saw the `e` in Medford but as I continue typing, it was deleted without me pressing backspace.

Woah, that is definitely not supposed to happen. I can't reproduce it for some reason, but I will look at this now and try to figure out what's going on. What browser are you using? Want to rule that out first.

**Comment by PaulJKim** [comment]
@cmaddox5 I'm using Firefox. And yeah I couldn't pick out a pattern. The only way I could reproduce it is just by constantly typing in various strings until it eventually happens maybe ever 4-7 tries or so. Could certainly be something unique to me as well.

File: assets/js/components/Dashboard/PermanentConfiguration/PlacesSearchBar.tsx [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+      </span>
+    );
+  };
+
+  // When pressing enter on a result instead of clicking the result, inputString is not cleared properly.
+  // As an item is selected, inputString will be set to the full name of the item.
+  // As the parent state changes, check if the inputString is the full item name or id.
+  // If it is, the item was just selected and the text box needs to be cleared.
+  useEffect(() => {
+    const place = places.find(
```
[the remark this answers is not in the document]
Comment (cmaddox5) [reply assets/js/components/Dashboard/PermanentConfiguration/PlacesSearchBar.tsx]: Good callout. Totally overlooked that. I renamed `handleSearch` to something more general and called that from in the hook.

**Comment by PaulJKim** [comment]
Ok so I'm still not 100% sure but it seems like the bug happens when and update to the results list is a little bit delayed while I'm still typing. Once the results list catches up and fully renders, it will sometimes set me back a character. It seems to only happen when I'm typing a little on the quicker side.

**Comment by cmaddox5** [comment]
Man, this is a tough one. I might need to pair with you to try and get an idea on what the problem is. It must have something to do with the package I'm using, but I can't reproduce it for some reason. Gonna try to hunt down the issue today and tomorrow and will Slack you next week if I still can't figure it out.

**Comment by cmaddox5** [comment]
I think this is good to go now. After making the code changes we discussed on the call, I had to tweak how the component is reset back to it's default state. Normally, maintaining our own state would be the solution but that was what was causing the bug you found. Now, I am keeping an integer in state, making that the `key` prop of the component, and flipping it between -1 and 1 in order to trick React to rerender the component. Not ideal as that should be the responsibility of the component, but that package hasn't had any activity in a while.

**Review by PaulJKim — APPROVED** [review]
This looks good to me!

**Pull request merged**